File: ts/src/perfcli/perfcli.ts
import axios from 'axios';

const commandLineArgs = require('command-line-args');
const B = 1000000000; // 1 Billion, no. of nanoseconds in a second
function hrtime2Seconds(hrtime: number[]): number {
  return (hrtime[0]*B + hrtime[1])/B
}
const optionDefinitions = [
	{ name: 'url', alias: 'u', type: String },
	{ name: 'method', alias: 'm', type: String, defaultValue: 'get' },
	{ name: 'threads', alias: 't', type: Number, defaultValue: 1 },
	{ name: 'times', alias: 'n', type: Number, defaultValue: 1 },
	{ name: 'minDelay', alias: 'x', type: Number, defaultValue: 100 },
	{ name: 'maxDelay', alias: 'X', type: Number, defaultValue: 1000 },
];
const options = commandLineArgs(optionDefinitions);

if (!options.url) {
	console.log('No url specified');
	process.exit(1);
}
if (options.maxDelay <= options.minDelay) {
	options.maxDelay = options.minDelay + 1
}
console.log('options:', options);
execute().catch(error => { console.log('Error:', error); });

async function execute() {
	const st = process.hrtime();
	for (let i = 0; i < options.threads; i++) {
		await clientSimulator(i);
	}
	const et = process.hrtime(st);
	console.log(`Total Elapsed Time: ${hrtime2Seconds(et)} secs`);
}

async function sleepMilliSecs(msecs: number) {
	return new Promise(resolve => setTimeout(resolve, msecs));
}

async function clientSimulator(index: number) {
	let [firstDuration, minDuration, maxDuration, totalDuration, avgDuration] = [0.0, 100.0, 0.0, 0.0, 0.0];
	const start = process.hrtime();
	for (let k = 0; k < options.times; k++) {
		const delay = options.minDelay + Math.floor(Math.random()*(options.maxDelay - options.minDelay));
		await sleepMilliSecs(delay);
		const st = process.hrtime();
		let et = null;
		try {
			await axios.get(options.url);
			et = process.hrtime(st);
		} catch (e) {
			console.log(`[${index}, ${k}] Error:`, e);
			continue;
		}
		let respDuration = hrtime2Seconds(et);
		console.log(`[${index}, ${k}] Response Time: ${respDuration} secs.`);
		if (k === 0) {
			firstDuration = respDuration;
		} else {
			totalDuration += respDuration
			if (respDuration < minDuration) {
				minDuration = respDuration;
			} else if (respDuration > maxDuration) {
				maxDuration = respDuration
			}
		}
	}
	if (options.times > 1) {
		avgDuration = totalDuration/(options.times - 1);
	}
	const elapsedDuration = hrtime2Seconds(process.hrtime(start));
	console.log(`[${index}] Times: ${options.times}, Elapsed: ${elapsedDuration.toFixed(3)} secs, ` +
		`First: ${firstDuration.toFixed(3)} secs, Min: ${minDuration.toFixed(3)} secs, ` +
		`Max: ${maxDuration.toFixed(3)} secs, Avg: ${avgDuration.toFixed(3)} secs`);
}

File: go/src/perfcli/perfcli.go
package main
import (
	"fmt"
	"flag"
	"io"
	"io/ioutil"
	"net/http"
	"os"
	"time"
	"math/rand"
)
var tr *http.Transport = &http.Transport{}
var client *http.Client = &http.Client{Transport: tr}
var url string = ""
var method string = "get"
var threads int = 1
var times int = 1
var minDelay int = 100
var maxDelay int = 1000
var closeConnection bool = false

func main() {
	flag.StringVar(&url, "url", "", "target url")
	flag.StringVar(&method, "method", "get", "HTTP method")
	flag.IntVar(&threads, "threads", 1, "number of threads to make requests")
	flag.IntVar(&times, "times", 1, "number of times to request for each thread")
	flag.IntVar(&minDelay, "minDelay", 100, "min time to wait before next request (ms)")
	flag.IntVar(&maxDelay, "maxDelay", 1000, "max time to wait before next request (ms)")
	flag.BoolVar(&closeConnection, "closeConnection", false, "Close connection? (default: false)")
	
	
	flag.Parse();
	if url == "" {
		fmt.Fprintf(os.Stderr, "args: no url specified\n")
		os.Exit(1)
	}
	if maxDelay <= minDelay {
		maxDelay = minDelay + 1
	}
	fmt.Printf("url            : %v\n", url)
	fmt.Printf("method         : %v\n", method)
	fmt.Printf("threads        : %v\n", threads)
	fmt.Printf("times          : %v\n", times)
	fmt.Printf("minDelay       : %v\n", minDelay)
	fmt.Printf("maxDelay       : %v\n", maxDelay)
	fmt.Printf("closeConnection: %v\n", closeConnection)
    
	start := time.Now()
	ch := make(chan string)
	for i := 0; i < threads; i++ {
		go clientSimulator(ch, i);
	}
	for i := 0; i < threads; i++ {
		fmt.Printf("%s\n", <-ch)
	}
	elapsedSecs := time.Since(start).Seconds()
	fmt.Printf("Total Elapsed Time: %.3f secs\n", elapsedSecs)
}

func clientSimulator(ch chan<- string, index int) {
	totalDuration := time.Duration(0)
	minDuration := time.Duration(1 * time.Hour)
	maxDuration := time.Duration(0)
	firstDuration := time.Duration(0)
	avgDuration := 0.0
	start := time.Now()
	for k := 0; k < times; k++ {
		delay := time.Duration(minDelay + rand.Intn(maxDelay - minDelay))
		time.Sleep(delay * time.Millisecond)
		
		reqStart := time.Now()
		resp, err := http.Get(url)
		if err != nil {
			fmt.Fprintf(os.Stderr, "clientSimulator: http.Get(%s) failed: %v\n", url, err)
			continue
		}
		if closeConnection {
			tr.CloseIdleConnections()
		} else {
			_, err = io.Copy(ioutil.Discard, resp.Body)
			if err != nil {
				fmt.Fprintf(os.Stderr, "clientSimulator: io.Copy() failed: %v\n", err)
				continue
			}
		}

		respDuration := time.Since(reqStart)
		if k == 0 {
			firstDuration = respDuration
		} else {
			totalDuration += respDuration
			switch {
				case respDuration < minDuration:
					minDuration = respDuration
				case respDuration > maxDuration:
					maxDuration = respDuration
			}
		}
		fmt.Printf("[%d, %d] Response Time: %.3f secs\n", index, k, respDuration.Seconds())
	}
	if times > 1 {
		avgDuration = totalDuration.Seconds()/(float64)(times - 1)
	}
	elapsedSecs := time.Since(start).Seconds()
	ch <- fmt.Sprintf("[%d] Times: %d, Elapsed: %.3f, First: %.3f, Min: %.3f, Max: %.3f, Avg: %.3f", 
		index, times, elapsedSecs, firstDuration.Seconds(), minDuration.Seconds(), maxDuration.Seconds(), avgDuration)
}
